src/app/server/DefaultTermsAndConditionsProvider.cpp

```diff
@@ -0,0 +1,300 @@
+    uint16_t acknowledgements        = 0;
+    uint16_t acknowledgementsVersion = 0;
+
+    chip::TLV::TLVReader tlvReader;
+    chip::TLV::TLVType tlvContainer;
+
+    uint8_t buffer[kEstimatedTlvBufferSize] = { 0 };
+    uint16_t bufferSize                     = sizeof(buffer);
+
+    const chip::StorageKeyName kStorageKey = chip::DefaultStorageKeyAllocator::TermsAndConditionsAcceptance();
+    CHIP_ERROR err                         = mStorageDelegate->SyncGetKeyValue(kStorageKey.KeyName(), &buffer, bufferSize);
+    if (CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND == err)
+    {
+        outTermsAndConditions.ClearValue();
+        return CHIP_NO_ERROR;
+    }
+
+    VerifyNoErrorOrReturnInternal(err);
+
+    tlvReader.Init(buffer, bufferSize);
+    VerifyNoErrorOrReturnInternal(tlvReader.Next(chip::TLV::kTLVType_Structure, chip::TLV::AnonymousTag()));
+    VerifyNoErrorOrReturnInternal(tlvReader.EnterContainer(tlvContainer));
+    VerifyNoErrorOrReturnInternal(tlvReader.Next(kSerializationVersionTag));
+    VerifyNoErrorOrReturnInternal(tlvReader.Get(serializationVersion));
+    VerifyNoErrorOrReturnInternal(tlvReader.Next(kAcceptedAcknowledgementsTag));
+    VerifyNoErrorOrReturnInternal(tlvReader.Get(acknowledgements));
+    VerifyNoErrorOrReturnInternal(tlvReader.Next(kAcceptedAcknowledgementsVersionTag));
+    VerifyNoErrorOrReturnInternal(tlvReader.Get(acknowledgementsVersion));
+    VerifyNoErrorOrReturnInternal(tlvReader.ExitContainer(tlvContainer));
+
+    // Only v1 serialization format is supported
+    VerifyOrReturnInternal(kSerializationVersion == serializationVersion);
+
+    outTermsAndConditions = Optional<TermsAndConditions>(TermsAndConditions{
+        .value   = acknowledgements,
+        .version = acknowledgementsVersion,
+    });
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR chip::app::DefaultTermsAndConditionsStorageDelegate::Set(const TermsAndConditions & inTermsAndConditions)
+{
+    uint8_t buffer[kEstimatedTlvBufferSize] = { 0 };
+    chip::TLV::TLVWriter tlvWriter;
+    chip::TLV::TLVType tlvContainer;
+
+    VerifyNotNullOrReturnUninitialized(mStorageDelegate);
+
+    tlvWriter.Init(buffer, sizeof(buffer));
+    VerifyNoErrorOrReturnInternal(tlvWriter.StartContainer(chip::TLV::AnonymousTag(), chip::TLV::kTLVType_Structure, tlvContainer));
+    VerifyNoErrorOrReturnInternal(tlvWriter.Put(kSerializationVersionTag, kSerializationVersion));
+    VerifyNoErrorOrReturnInternal(tlvWriter.Put(kAcceptedAcknowledgementsTag, inTermsAndConditions.value));
+    VerifyNoErrorOrReturnInternal(tlvWriter.Put(kAcceptedAcknowledgementsVersionTag, inTermsAndConditions.version));
+    VerifyNoErrorOrReturnInternal(tlvWriter.EndContainer(tlvContainer));
+    VerifyNoErrorOrReturnInternal(tlvWriter.Finalize());
+
+    uint32_t lengthWritten = tlvWriter.GetLengthWritten();
+    VerifyOrReturnInternal(CanCastTo<uint16_t>(lengthWritten));
+    VerifyOrReturnInternal(lengthWritten <= kEstimatedTlvBufferSize);
```

Comment:
I mean, if this is false you already had a buffer overrun and smashed your stack and probably can't trust anything....  So not sure what this check is really getting you.

Reply:
OK, I removed the check